### PR TITLE
parse services for jsdoc and present rpc-manger with the metadata Fixes #1605

### LINKED ISSUE #1605: services and RPCs should support annotations for descriptions, type info, etc
- [x] pick a js parser and doc standard. ( doctrine & google closure ? )
- [x] add JsDoc style comments to RPCs as a test
- [x] extract, parse the comments and attach to the RPC function

### PR DESCRIPTION
**Pull request by hamidzr**
- scans services source codes for jsdoc blocks
- parses jsdocs
- finds the associated RPC with the parsed jsdoc
- provides this meta data to the rpc-manager
- + when building the RPCs, grabs the argument names from jsdoc if available

### CODE REVIEW

File: src/server/rpc/jsdoc-extractor.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                indent = 0;
+            }
+        }
+
+        number += 1;
+        return result;
+    };
+}
+
+const FILE= process.argv[2];
```
Comment (brollb) [inline src/server/rpc/jsdoc-extractor.js]: should only run this way if running from the cli, right? Otherwise, I would just remove this code and add a utility script that imports this and passes the file name.

File: src/server/rpc/procedures/geolocation/geolocation.js [hunk 2]
```diff
@@ -58,7 +58,11 @@ if(!process.env.GOOGLE_GEOCODING_API) {
     };
 
 
-    // geocode an address and send back the details
+    /**
+     * geolocate and address and send back the details
```
Comment (brollb) [inline src/server/rpc/procedures/geolocation/geolocation.js]: This description needs some work since these will eventually be sent back to the users: 
- capitalize the first letter
- I think "and" is a typo
-  what are the details? 

I would try something else like "Get a ___ from an address"

File: test/server/rpc/jsdoc-extractor.spec.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    GoogleMap.doStuff = (asdf) =>`;
+
+        let testLines = testText.split('\n');
+
+        it('should support multiline', () => {
+            assert.deepEqual(jp._findFn(testLines[2]), 'doStuff');
+        });
+
+        it('should find let fn = ()', () => {
+            let line = 'let reverseGeocode = (lat, lon, response, query)=>{';
```
Comment (brollb) [inline test/server/rpc/jsdoc-extractor.spec.js]: what about an arrow function with only one argument and no parenthesis?

File: test/server/rpc/jsdoc-extractor.spec.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            assert.deepEqual(jp._findFn(line), 'reverseGeocode');
+        });
+
+        it('should fiind obj.ojb = function', () => {
```
Comment (brollb) [inline test/server/rpc/jsdoc-extractor.spec.js]: two typos in this line

File: src/server/rpc/rpc-manager.js [hunk 5]
```diff
@@ -55,6 +60,8 @@ RPCManager.prototype.loadRPCs = function() {
         })
         .map(pair => {
             let [name, RPCConstructor] = pair;
+            // console.log(RPCConstructor._docs);
```
Comment (brollb) [inline src/server/rpc/rpc-manager.js]: I would remove these comments

**Comment by brollb** [comment]
It looks like a good start! After looking at the PR, I think it might be good to include something that actually uses the parsed docs somehow. Something like getting the argument names from the doc if it exists and falling back on the current method. Does that make sense?

Also, a couple little fixes:
- [ ] the tests are failing
- [ ] the PR should be given a better name
- [ ] I would check out the codeclimate issues

**Comment by hamidzr** [comment]
yeah I didn't mean to make the PR at this point, it was misclick.

**Comment by brollb** [comment]
cool, no worries!

File: test/server/rpc/jsdoc-extractor.spec.js (hunk 3 above)
Comment (hamidzr) [reply test/server/rpc/jsdoc-extractor.spec.js]: I added that. let me know if there is any other cases

File: src/server/rpc/procedures/geolocation/geolocation.js [hunk 6]
```diff
@@ -58,7 +58,12 @@ if(!process.env.GOOGLE_GEOCODING_API) {
     };
 
 
-    // geocode an address and send back the details
+    /**
+     * Geolocates the address and sends back the coordinates
+     * @param {string} address target address
+     * @returns {structuredData}
```
Comment (brollb) [inline src/server/rpc/procedures/geolocation/geolocation.js]: I would just say the javascript form for these things here (ie, `Object`)

Comment (brollb) [reply src/server/rpc/procedures/geolocation/geolocation.js]: The idea is that the RPCs should be able to be written just as regular javascript with minimal additional knowledge required about the rest of the system and then netsblox will just automatically convert between JavaScript and NetsBlox types as needed.

File: src/client/blocks-ext.js [hunk 7]
```diff
@@ -257,7 +257,7 @@ RPCInputSlotMorph.prototype.methodSignature = function () {
         this.fieldsFor = JSON.parse(this.getURL('/rpc/' + rpc));
 
         // grab the list of deprecated methods
-        let deprecatedMethods = this.fieldsFor._compability.deprecatedMethods || [];
+        var deprecatedMethods = this.fieldsFor._compability.deprecatedMethods || [];
```
Comment (brollb) [inline src/client/blocks-ext.js]: This should be a different PR

File: src/server/rpc/jsdoc-extractor.js [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        let linesToSearch = lines.slice(block.endLine, block.endLine + searchScope);
+        let fnName;
+        // if @name is set jsut use that and save a few cycles
+        if (block.parsed.tags.some(tag => tag.title === 'name')) {
+            fnName  = block.parsed.tags.find(tag => tag.title === 'name').name;
+            logger.info('fn name set through @name', fnName);
+        } else {
+            fnName = findFn(linesToSearch);
+            if (!fnName){
+                logger.error(`can't associate ${block.lines} with any function. # Fix it at line ${block.beginLine}, column ${block.column}`);
```
Comment (brollb) [inline src/server/rpc/jsdoc-extractor.js]: I would change this to a warning instead of an error. Also, I would print the actual comment that you cannot parse instead of just the line row and column.

File: src/server/rpc/jsdoc-extractor.js [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    blocks = blocks.map(block => {
+        let src = block.lines.join('\n');
+        block.parsed = doctrine.parse(src, {unwrap: true});
+        return block;
+    });
+
+    blocks = blocks.filter(block => {
+        let linesToSearch = lines.slice(block.endLine, block.endLine + searchScope);
+        let fnName;
+        // if @name is set jsut use that and save a few cycles
```
Comment (brollb) [inline src/server/rpc/jsdoc-extractor.js]: jsut -> just

File: src/server/rpc/jsdoc-extractor.js [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        if (block.parsed.tags.some(tag => tag.title === 'name')) {
+            fnName  = block.parsed.tags.find(tag => tag.title === 'name').name;
```
Comment (brollb) [inline src/server/rpc/jsdoc-extractor.js]: It isn't a huge deal but you could combine the top two so you aren't searching for the same thing twice.
```
let nameTag = block.parsed.tags.find(tag => tag.title === 'name');
if (nameTag) {
    fnName = nameTag.name;
```

**Comment by hamidzr** [comment]
failing test looks like to be a problem with Travis's config

File: src/server/rpc/procedures/geolocation/geolocation.js [hunk 11]
```diff
@@ -58,7 +58,12 @@ if(!process.env.GOOGLE_GEOCODING_API) {
     };
 
 
-    // geocode an address and send back the details
+    /**
+     * Geolocates the address and sends back the coordinates
```
Comment (brollb) [inline src/server/rpc/procedures/geolocation/geolocation.js]: I am not sure about the wording... We should probably come up with guidelines for these... "returns" might be better than "sends back"

Comment (brollb) [reply src/server/rpc/procedures/geolocation/geolocation.js]: It isn't necessary to come up with guidelines for this PR but it is something that should be done

File: src/server/rpc/jsdoc-extractor.js (hunk 8 above)
Comment (hamidzr) [reply src/server/rpc/jsdoc-extractor.js]: .lines is the actual lines

**Comment by hamidzr** [comment]
>> One thing I would consider is generating something more domain specific from the jsdoc immediately
what information do you think we should have from what doctrine provides

**Comment by brollb** [comment]
The last example was supposed to be something like this:
```
{
    name: 'geolocate',
    description: 'Get location information about a given address',
    arguments: [
        {
            name: 'address',
            type: 'string',
            description: 'An address like "1025 W 4th Street, Nashville, TN'
        }
    ]
}
```

**Comment by hamidzr** [comment]
https://github.com/NetsBlox/NetsBlox/issues/1605#issuecomment-337278424 should be resolved before merging

File: test/server/rpc/jsdoc-extractor.spec.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+describe('jsdoc-extractor', () => {
+
+    let comment = `
+    /**
+     * this is the description
+     * next line of description
+     * @param {string} address target address
+     * @param {number} limit the results limit
+     * @param {Object} options
```
Comment (brollb) [inline test/server/rpc/jsdoc-extractor.spec.js]: A minor point but it would be nice to follow one convention for the types everywhere (even in test cases). Specifically, whether or not we are capitalizing the first word of the type like `Number` (the fn in JavaScript) or `number` (the output of `typeof 5` in JavaScript).

File: src/server/rpc/procedures/geolocation/geolocation.js [hunk 13]
```diff
@@ -58,7 +58,12 @@ if(!process.env.GOOGLE_GEOCODING_API) {
     };
 
 
-    // geocode an address and send back the details
+    /**
+     * Geolocates the address and returns the coordinates
+     * @param {string} address target address
```
Comment (brollb) [inline src/server/rpc/procedures/geolocation/geolocation.js]: same as later comment about consistency among the type naming

**Comment by hamidzr** [comment]
my intent was to keep them all lower case like typeof.

Hamid Zare

**Comment by brollb** [comment]
It might be nice to use uppercase instead so it is more obvious with future types. For example, with latitude, longitude types, it might be nicer to say:
```
@param {Latitude} latitude
```
rather than
```
@param {latitude} latitude
```
Additionally, this should promote consistency when used programmatically as it would be nice to not have to change case but still maintain consistency with the existing constructors for the types: Number, String, etc.

Thoughts?

**Comment by hamidzr** [comment]
agreed
